### PR TITLE
Add "Edit question" option to the dashcard menu on the dashboard

### PR DESCRIPTION
**Pull request by ranquild**
Epic https://github.com/metabase/metabase/issues/29648

Add "Edit question" option to the dashcard menu on the dashboard.

Changes:
- Extracts the popover from `QueryDownloadWidget`  into a separate `QueryDownloadPopover` component
- Extracts download logic from `QueryDownloadWidget`  to redux actions
- Creates a new `DashCardMenu` component that uses `QueryDownloadPopover` and download actions

Screenshot 2023-04-12 at 13 43 21
Screenshot 2023-04-12 at 13 43 28

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/29937)

### CODE REVIEW

File: frontend/src/metabase/lib/urls/questions.ts [hunk 1]
```diff
@@ -74,7 +80,9 @@ export function question(
     path = appendSlug(path, slugg(name));
   }
 
-  if (objectId) {
+  if (mode === "notebook") {
```
Comment (ranquild) [inline frontend/src/metabase/lib/urls/questions.ts:83]: Previously it wasn't possible to generate a url for the notebook

File: frontend/src/metabase/query_builder/actions/downloading.ts [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+  visualizationSettings?: VisualizationSettings;
+}
+
+interface DownloadQueryResultsParams {
+  method: string;
+  url: string;
+  params: URLSearchParams;
+}
+
+export const downloadQueryResults =
```
Comment (ranquild) [inline frontend/src/metabase/query_builder/actions/downloading.ts:32]: Extracted from `QueryDownloadWidget.jsx`

**Review by mazameli — COMMENTED** [review]
Ideally, for SQL/native questions the Edit Question action would go to the question with the SQL editor already open/expanded. Would that be hard to do?

**Review by WiNloSt — COMMENTED** [review]
This is the first round of review, I'll continue tomorrow.

But visually I only found 1 place we could fix. The new entity button doesn't align with the question tooltip (if the description is set)

#### Before



#### This PR



> Reviewed 10 of 27 files at r1, 1 of 3 files at r2, 1 of 5 files at r3, 4 of 6 files at r4, 2 of 4 files at r5.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29937#-:-NSvQkWz0XDqZE43_jJY:b-v1fl60)* status: 18 of 35 files reviewed, 7 unresolved discussions (waiting on @ranquild)

___
*[`e2e/test/scenarios/dashboard/dashboard.cy.spec.js` line 580 at r5](https://reviewable.io/reviews/metabase/metabase/29937#-NSv8Ua_DlCSt4D5mFD0:-NSv8Ua_DlCSt4D5mFD1:b-9l6201) ([raw file](https://github.com/metabase/metabase/blob/6e41a406e94364cab193ec4125be934efd24a0b1/e2e/test/scenarios/dashboard/dashboard.cy.spec.js#L580)):*
> ```JavaScript
>     getDashboardCardMenu().click();
>     popover().findByText("Edit question").click();
>     cy.findByText("Visualize").should("be.visible");
> ```

could you do `cy.button("Visualize").should("be.visible")`? At least it's clearer we're looking for this button.

___
*[`frontend/src/metabase/icon_paths.ts` line 162 at r5](https://reviewable.io/reviews/metabase/metabase/29937#-NSv9SuY7mpvl4gGB1_y:-NSv9SuZE9msre6LwZXM:b-z0y990) ([raw file](https://github.com/metabase/metabase/blob/6e41a406e94364cab193ec4125be934efd24a0b1/frontend/src/metabase/icon_paths.ts#L162)):*
> ```XML
>     "M29,10.1052632 L29,28.8325291 C29,30.581875 27.5842615,32 25.8337327,32 L7.16626728,32 C5.41758615,32 4,30.5837102 4,28.8441405 L4,3.15585953 C4,1.41292644 5.42339685,9.39605581e-15 7.15970573,8.42009882e-15 L20.713352,8.01767853e-16 L20.713352,8.42105263 L22.3846872,8.42105263 L22.3846872,0.310375032 L28.7849894,8.42105263 L20.713352,8.42105263 L20.713352,10.1052632 L29,10.1052632 Z M7.3426704,12.8000006 L25.7273576,12.8000006 L25.7273576,14.4842112 L7.3426704,14.4842112 L7.3426704,12.8000006 Z M7.3426704,17.3473687 L25.7273576,17.3473687 L25.7273576,19.0315793 L7.3426704,19.0315793 L7.3426704,17.3473687 Z M7.3426704,21.8947352 L25.7273576,21.8947352 L25.7273576,23.5789458 L7.3426704,23.5789458 L7.3426704,21.8947352 Z M7.43137255,26.2736849 L16.535014,26.2736849 L16.535014,27.9578954 L7.43137255,27.9578954 L7.43137255,26.2736849 Z",
>   download:
>     "M 16 7 c 4.4702 0 8.1988 3.1778 9.0486 7.3982 C 28.9255 14.6209 32 17.8363 32 21.7692 c 0 4.0782 -3.3065 7.3846 -7.3846 7.3846 H 7.3846 C 3.3065 29.1538 0 25.8474 0 21.7692 c 0 -3.9329 3.0751 -7.1483 6.952 -7.3723 C 7.8012 10.1778 11.5298 7 16 7 z m 0 3.6923 a 1.8462 1.8462 90 0 0 -1.8462 1.8462 v 8.3994 l -3.2197 -3.3914 l -0.08 -0.0788 a 1.4825 1.4825 90 0 0 -2.096 0.0788 c -0.6006 0.6326 -0.6006 1.6591 0 2.2917 l 6.1538 6.4818 l 0.08 0.0794 a 1.4825 1.4825 90 0 0 2.096 -0.0794 l 6.1538 -6.4818 l 0.0751 -0.0849 c 0.5243 -0.6363 0.4997 -1.6012 -0.0751 -2.2068 l -0.08 -0.0788 a 1.4825 1.4825 90 0 0 -2.096 0.0788 L 17.8462 20.9378 V 12.5385 a 1.8462 1.8462 90 0 0 -1.8462 -1.8462 z",
> ```

I honestly couldn't see any visual difference unless maybe a 1px shift in the icon

#### Before


#### After


I wonder if this change is necessary.
___
*[`frontend/src/metabase/dashboard/components/DashCard/DashCardVisualization.tsx` line 183 at r5](https://reviewable.io/reviews/metabase/metabase/29937#-NSvMgju1MFuah7IFopl:-NSvMgju1MFuah7IFopm:b6z4f3n) ([raw file](https://github.com/metabase/metabase/blob/6e41a406e94364cab193ec4125be934efd24a0b1/frontend/src/metabase/dashboard/components/DashCard/DashCardVisualization.tsx#L183)):*
> ```XML
>   const renderActionButtons = useCallback(() => {
>     const question = new Question(dashcard.card, metadata);
>     const mainSeries = series[0] as unknown as Dataset;
> ```

Is the reason we need to cast this to `Dataset` because the types are mismatched? And what do you think we should do should avoid casting in the future?
___
*[`frontend/src/metabase/dashboard/components/DashCard/DashCardMenu/DashCardMenu.tsx` line 115 at r5](https://reviewable.io/reviews/metabase/metabase/29937#-NSv5Xak-FZRoECf2Qw8:-NSv5Xak-FZRoECf2Qw9:b81pd1l) ([raw file](https://github.com/metabase/metabase/blob/6e41a406e94364cab193ec4125be934efd24a0b1/frontend/src/metabase/dashboard/components/DashCard/DashCardMenu/DashCardMenu.tsx#L115)):*
> ```XML
>           triggerProps={{
>             className: !open ? "hover-child hover-child--smooth" : undefined,
>             "data-testid": "dashcard-menu",
> ```

Could we use `aria-label` here instead? So we could query this menu by `cy.button("Dashcard menu")`
___
*[`frontend/src/metabase/lib/dom.js` line 487 at r5](https://reviewable.io/reviews/metabase/metabase/29937#-NSvQ18G5LJr-fMpwhYB:-NSvQ18G5LJr-fMpwhYC:b-uxeu97) ([raw file](https://github.com/metabase/metabase/blob/6e41a406e94364cab193ec4125be934efd24a0b1/frontend/src/metabase/lib/dom.js#L487)):*
> ```JavaScript
> 
> export function isReducedMotionPreferred() {
>   const mediaQuery = window.matchMedia?.("(prefers-reduced-motion: reduce)");
> ```

Why does this need to be changed?

**Comment by ranquild** [comment]
I hoped that we could use the standard EntityMenu trigger for this but seems like not. I've made it use just the `Icon` as it was before.

**Comment by ranquild** [comment]
*[`e2e/test/scenarios/dashboard/dashboard.cy.spec.js` line 580 at r5](https://reviewable.io/reviews/metabase/metabase/29937#-NSv8Ua_DlCSt4D5mFD0:-NSzpTur3pf8xloANv4R:b-886gsx) ([raw file](https://github.com/metabase/metabase/blob/6e41a406e94364cab193ec4125be934efd24a0b1/e2e/test/scenarios/dashboard/dashboard.cy.spec.js#L580)):*
<details><summary><i>Previously, WiNloSt (Mahatthana (Kelvin) Nomsawadi) wrote…</i></summary><blockquote>

could you do `cy.button("Visualize").should("be.visible")`? At least it's clearer we're looking for this button.


</blockquote></details>

Fixed

**Comment by ranquild** [comment]
*[`frontend/src/metabase/components/EntityMenuItem/EntityMenuItem.tsx` line 40 at r8](https://reviewable.io/reviews/metabase/metabase/29937#-NSzpWYR5vPlWiKC1VYD:-NSzpWYR5vPlWiKC1VYE:b-x2xsum) ([raw file](https://github.com/metabase/metabase/blob/836649241c79aff82f644b1f42b357c3f1968b57/frontend/src/metabase/components/EntityMenuItem/EntityMenuItem.tsx#L40)):*
> ```XML
>   const content = (
>     <MenuItemContent disabled={disabled}>
>       {icon && <MenuItemIcon name={icon} size={16} />}
> ```

@WiNloSt I've made this change and reverted changes to `icon_paths`. The problem I was trying to solve with these changes:

Before


After

**Comment by ranquild** [comment]
*[`frontend/src/metabase/dashboard/components/DashCard/DashCardVisualization.tsx` line 183 at r5](https://reviewable.io/reviews/metabase/metabase/29937#-NSvMgju1MFuah7IFopl:-NSzpmZTAOvKiKfzyNMO:b-nwlkm1) ([raw file](https://github.com/metabase/metabase/blob/6e41a406e94364cab193ec4125be934efd24a0b1/frontend/src/metabase/dashboard/components/DashCard/DashCardVisualization.tsx#L183)):*
<details><summary><i>Previously, WiNloSt (Mahatthana (Kelvin) Nomsawadi) wrote…</i></summary><blockquote>

Is the reason we need to cast this to `Dataset` because the types are mismatched? And what do you think we should do should avoid casting in the future?
</blockquote></details>

Yeah the problem is that `series` here uses types from `metabase-types/types` while `Dataset` is from `metabase-types/api`. They are the same thing but have different properties defined. Ideally we should use `Dataset` everywhere but it would require non-trivial type changes, and I thought that it's out of scope for this PR.

**Review by ranquild — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29937)* status: 16 of 36 files reviewed, 8 unresolved discussions (waiting on @ranquild and @WiNloSt)

___
*[`frontend/src/metabase/dashboard/components/DashCard/DashCardMenu/DashCardMenu.tsx` line 115 at r5](https://reviewable.io/reviews/metabase/metabase/29937#-NSv5Xak-FZRoECf2Qw8:-NSzq7zZ23UJlkVG0TeX:b-39u49x) ([raw file](https://github.com/metabase/metabase/blob/6e41a406e94364cab193ec4125be934efd24a0b1/frontend/src/metabase/dashboard/components/DashCard/DashCardMenu/DashCardMenu.tsx#L115)):*
<details><summary><i>Previously, WiNloSt (Mahatthana (Kelvin) Nomsawadi) wrote…</i></summary><blockquote>

Could we use `aria-label` here instead? So we could query this menu by `cy.button("Dashcard menu")`
</blockquote></details>

My reasoning was that the current implementation used the test id, so I didn't change this.
___
*[`frontend/src/metabase/lib/dom.js` line 487 at r5](https://reviewable.io/reviews/metabase/metabase/29937#-NSvQ18G5LJr-fMpwhYB:-NSzqFga0mNO3mAY_Wvk:b-z41wpw) ([raw file](https://github.com/metabase/metabase/blob/6e41a406e94364cab193ec4125be934efd24a0b1/frontend/src/metabase/lib/dom.js#L487)):*
<details><summary><i>Previously, WiNloSt (Mahatthana (Kelvin) Nomsawadi) wrote…</i></summary><blockquote>

Why does this need to be changed?
</blockquote></details>

Because `matchMedia` is not implemented in `js-dom`
___
*[`frontend/src/metabase/icon_paths.ts` line 162 at r5](https://reviewable.io/reviews/metabase/metabase/29937#-NSv9SuY7mpvl4gGB1_y:-NSzqO9E75DUQQGYQkab:b-9ktkd2) ([raw file](https://github.com/metabase/metabase/blob/6e41a406e94364cab193ec4125be934efd24a0b1/frontend/src/metabase/icon_paths.ts#L162)):*
<details><summary><i>Previously, WiNloSt (Mahatthana (Kelvin) Nomsawadi) wrote…</i></summary><blockquote>

I honestly couldn't see any visual difference unless maybe a 1px shift in the icon

#### Before


#### After


I wonder if this change is necessary.
</blockquote></details>

Made a different, hopefully more straightforward fix in `EntityMenuItem`, please take a look

**Review by WiNloSt — APPROVED** [review]
This is a huge change. I've added comments and questions. But nothing blocking you from shipping this.

> Reviewed 11 of 27 files at r1, 1 of 5 files at r3, 2 of 6 files at r4, 2 of 4 files at r5, 3 of 4 files at r8, 1 of 1 files at r9, 1 of 1 files at r10, 2 of 2 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/29937#-:-NT-dEmKBUx-6b2Cz3Nc:bclxsm0)* status: all files reviewed, 10 unresolved discussions (waiting on @ranquild)

___
*[`frontend/src/metabase/dashboard/components/DashCard/DashCardMenu/DashCardMenu.tsx` line 115 at r5](https://reviewable.io/reviews/metabase/metabase/29937#-NSv5Xak-FZRoECf2Qw8:-NT-QsBB4RwHbCtSZwY_:b-iq3mgj) ([raw file](https://github.com/metabase/metabase/blob/6e41a406e94364cab193ec4125be934efd24a0b1/frontend/src/metabase/dashboard/components/DashCard/DashCardMenu/DashCardMenu.tsx#L115)):*
<details><summary><i>Previously, ranquild (Alexander Polyankin) wrote…</i></summary><blockquote>

My reasoning was that the current implementation used the test id, so I didn't change this.
</blockquote></details>

sounds reasonable :thumbsup: 
___
*[`frontend/src/metabase/query_builder/actions/downloading.ts` line 161 at r11](https://reviewable.io/reviews/metabase/metabase/29937#-NT-Uk_45aBxTBNiOiR3:-NT-Uk_45aBxTBNiOiR4:bbantr2) ([raw file](https://github.com/metabase/metabase/blob/d98b3619345005c26171ac2666f401a904fc4126/frontend/src/metabase/query_builder/actions/downloading.ts#L161)):*
> ```XML
> const openSaveDialog = (fileName: string, fileContent: Blob) => {
>   const url = URL.createObjectURL(fileContent);
>   const link = document.createElement(`a`);
> ```

nit: This is probably from the refactoring, but we should use plain quotes for static strings, but feel free to skip if you don't want to burn another CI cycles.
___
*[`frontend/src/metabase/query_builder/actions/downloading.ts` line 163 at r11](https://reviewable.io/reviews/metabase/metabase/29937#-NT-V-8q8b0-2gvx_XIh:-NT-V-8q8b0-2gvx_XIi:b-y4hpd0) ([raw file](https://github.com/metabase/metabase/blob/d98b3619345005c26171ac2666f401a904fc4126/frontend/src/metabase/query_builder/actions/downloading.ts#L163)):*
> ```XML
>   const link = document.createElement(`a`);
>   link.href = url;
>   link.setAttribute(`download`, fileName);
> ```

nit: Same here. These should be plain quotes
___
*[`frontend/src/metabase/query_builder/components/QueryDownloadPopover/QueryDownloadPopover.tsx` line 33 at r11](https://reviewable.io/reviews/metabase/metabase/29937#-NT-XCr05ma7QwlTj_vB:-NT-XCr05ma7QwlTj_vC:b-qc4ahm) ([raw file](https://github.com/metabase/metabase/blob/d98b3619345005c26171ac2666f401a904fc4126/frontend/src/metabase/query_builder/components/QueryDownloadPopover/QueryDownloadPopover.tsx#L33)):*
> ```XML
> type QueryDownloadPopoverProps = OwnProps & StateProps;
> 
> const mapStateToProps = (
> ```

I don't think I've seen `mapStateToProps` that doesn't use `state`. It feels like the intention of having `mapStateToProps` is to get certain states from the redux store, not for encapsulating logic.

Do you think we should just compute these values inside the component body instead? If you want to colocate these logics they could be put inside a new hook maybe.
___
*[`frontend/src/metabase/query_builder/components/QueryDownloadWidget/QueryDownloadWidget.tsx` line 65 at r11](https://reviewable.io/reviews/metabase/metabase/29937#-NT-bo2J9_xvM9Oi1cKP:-NT-bo2J9_xvM9Oi1cKQ:b9bl0i4) ([raw file](https://github.com/metabase/metabase/blob/d98b3619345005c26171ac2666f401a904fc4126/frontend/src/metabase/query_builder/components/QueryDownloadWidget/QueryDownloadWidget.tsx#L65)):*
> ```XML
>       triggerContent={
>         loading ? (
>           <Tooltip tooltip={t`Downloading…`}>
> ```

Can you help check what's wrong here? The new loading indicator slightly shifts from the previous position.

#### On stats


#### This PR

___
*[`frontend/src/metabase/query_builder/components/QueryDownloadWidget/QueryDownloadWidget.unit.spec.tsx` line 57 at r11](https://reviewable.io/reviews/metabase/metabase/29937#-NT-cPEyA6wkFEBt0qRL:-NT-cPEyA6wkFEBt0qRM:b-hxqxws) ([raw file](https://github.com/metabase/metabase/blob/d98b3619345005c26171ac2666f401a904fc4126/frontend/src/metabase/query_builder/components/QueryDownloadWidget/QueryDownloadWidget.unit.spec.tsx#L57)):*
> ```XML
>     userEvent.click(getIcon("download"));
> 
>     expect(await screen.findByText(/Download/)).toBeInTheDocument();
> ```

Can this be just a simple string assertion? So we know what exactly we want to see. e.g. this is the popover after clicking the download button, maybe we can assert `Download full results`. What do you think?

**Pull request merged**